rmw_fastrtps_cpp/src/rmw_client.cpp

```diff
@@ -120,7 +130,11 @@ rmw_create_client(
       participant, request_type_name.c_str(),
       reinterpret_cast<TopicDataType **>(&info->request_type_support_)))
   {
-    info->request_type_support_ = new RequestTypeSupport_cpp(service_members);
+    info->request_type_support_ = new (std::nothrow) RequestTypeSupport_cpp(service_members);
```

Comment:
Ha, this is interesting. You're spot on about the race, but alas these functions are not expected to be thread-safe to begin with (see https://github.com/ros2/rmw/pull/276#discussion_r492971067). 

I see much discussion incoming. I'll add it to the backlog.

Reply:
> Ha, this is interesting. You're spot on about the race, but alas these functions are not expected to be thread-safe to begin with (see ros2/rmw#276 (comment)).

So, the node-object doesn't have to be thread safe, we're mutexing all the accesses to `rcl_node_t` in `rclcpp`/`rclpy`.
The problem is that the function isn't concurrent at all (the domain participant factory is a singleton), and when you compose different nodes in the same process this function calls can randomly break.

I never have that problem with fastrtps, but that was a problem in connext. Of course, it was almost impossible to reproduce.

Reply:
> So, the node-object doesn't have to be thread safe, we're mutexing all the accesses to rcl_node_t in rclcpp/rclpy.

I'm surprised we're actually not doing that (in `rclcpp` at least)

Reply:
> The problem is that the function isn't concurrent at all (the domain participant factory is a singleton), and when you compose different nodes in the same process this function calls can randomly break.

We actually would need one mutex per context to avoid this race https://github.com/eProsima/Fast-DDS/blob/724fa5741d569eff6d5633b6eb587238df97604d/src/cpp/fastrtps_deprecated/Domain.cpp#L309-L316.

Reply:
> The problem is that the function isn't concurrent at all (the domain participant factory is a singleton), and when you compose different nodes in the same process this function calls can randomly break.

Hmm I see. And I agree, non-reentrant functions are a problem. I'll open an issue.